### PR TITLE
8302687 Implement interfaces and shared code for announcement feature

### PR DESCRIPTION
**Pull request by savoptik**
This enhancement covers basic API and shared code that should be implemented for the Accessibility Announcement feature.

CSR [JDK-8304499](https://bugs.openjdk.org/browse/JDK-8304499 "bugs.openjdk.org")

@mrserb @prrace @azuev-java please review

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302687](https://bugs.openjdk.org/browse/JDK-8302687): Implement interfaces and shared code for announcement feature (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13001/head:pull/13001` \
`$ git checkout pull/13001`

Update a local copy of the PR: \
`$ git checkout pull/13001` \
`$ git pull https://git.openjdk.org/jdk.git pull/13001/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13001`

View PR using the GUI difftool: \
`$ git pr show -t 13001`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13001.diff">https://git.openjdk.org/jdk/pull/13001.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13001#issuecomment-1466579889)

### CODE REVIEW

File: src/java.desktop/share/classes/sun/swing/AccessibleAnnounceProvider.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+  * This class provides the ability to speak a given string using screen readers.
+ */
+public class AccessibleAnnounceProvider {
+
+    /**
+     * messages do not interrupt the current speech, they are spoken after the screen reader has spoken the current phrase
+     */
+    @Native private static final int ANNOUNCE_WITHOUT_INTERRUPTING_CURRENT_OUTPUT = AccessibleAnnouncer.ANNOUNCE_WITHOUT_INTERRUPTING_CURRENT_OUTPUT;
```
[the remark this answers is not in the document]
Comment (savoptik) [reply src/java.desktop/share/classes/sun/swing/AccessibleAnnounceProvider.java:40]: Done.

**Comment by savoptik** [comment]
CSR [JDK-8304085](https://bugs.openjdk.org/browse/JDK-8304085 "bugs.openjdk.org")

File: test/jdk/java/awt/a11y/AccessibleAnnouncerTest.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            }
+        });
+
+        frame.setLayout(new FlowLayout());
+        frame.add(button);
+        exceptionString = "Accessible announcer priority test failed!";
+        super.createUI(frame, "Accessible Anouncer test");
+    }
+
+    public static void main(String[] args)  throws Exception {
```
Comment (turbanoff) [inline test/jdk/java/awt/a11y/AccessibleAnnouncerTest.java:127]:
```suggestion
    public static void main(String[] args) throws Exception {
```

Comment (savoptik) [reply test/jdk/java/awt/a11y/AccessibleAnnouncerTest.java]: Done

**Comment by savoptik** [comment]
Do not close this pull request.
@mrserb @prrace @azuev-java please review

File: src/java.desktop/share/classes/javax/accessibility/AccessibleAnnouncer.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    /**
+     * This method makes an announcement with the specified priority from an accessible to which the announcing relates
+     *
+     * @param a      an accessible to which the announcing relates
+     * @param str      string for announcing
+     * @param priority priority for announcing
+     */
+    void announce(Accessible a, final String str, final int priority);
+
+}
```
Comment (azuev-java) [inline src/java.desktop/share/classes/javax/accessibility/AccessibleAnnouncer.java:55]: Minor: no newline at the last line of the file. Not an error just a cosmetic.

**Review by azuev-java — COMMENTED** [review]
Since there is a new interface and a default implementation for it the CSR process should be invoked. See https://wiki.openjdk.org/display/csr/CSR+FAQs

Q: What sort of changes require CSR review?
A: Any change to a JDK interface meant to be used outside of the JDK itself requires CSR review. In this context "interface" isn't limited to the Java programing language definition of an interface, but encompasses the broader concept of a protocol between the JDK and users of the JDK. Examples of interfaces by this definition include:

- Changes to public and exported APIs in java.* and javax.* packages.

And that falls well within that clause.

**Comment by savoptik** [comment]
You don't need to close this pull request.

**Comment by savoptik** [comment]
You don't have to close anything.

**Comment by savoptik** [comment]
You don't need to close this pull request.

**Comment by prrace** [comment]
There's a really  tree of JBS issues related to this topic and I see a process problem.

The bug for this PR is https://bugs.openjdk.org/browse/JDK-8302687

But the CSR in this PR is listed as https://bugs.openjdk.org/browse/JDK-8304499
but that is actually the CSR for https://bugs.openjdk.org/browse/JDK-8302352
which is an umbrella bug.

So this is wrong.
The CSR must be the CSR for the bug in the PR, ie for something you will actually push!
Else skara and everyone else will get confused.

Then there's the overall question as to how important, or appropriate is this API ?
Seems like its not really an A11Y API, its an arbitrary text-to-speech API.
I think of the A11Y APIs as being tied to the UI and making it accessible, not
providing some other way of communicating something which isn't even in the UI.
Put it this way, if I am NOT using an AT to provide speech, how are you communicating
whatever it is to the user - something changes in the UI, right ?
So then the A11Y API will be able to report that already so why do you need this ?
Show me some equivalent cases and uses in platform ATs APIs

**Comment by savoptik** [comment]
@prrace
"There's a really tree of JBS issues related to this topic and I see a process problem.
The bug for this PR is https://bugs.openjdk.org/browse/JDK-8302687
But the CSR in this PR is listed as https://bugs.openjdk.org/browse/JDK-8304499
but that is actually the CSR for https://bugs.openjdk.org/browse/JDK-8302352
which is an umbrella bug.
So this is wrong.
The CSR must be the CSR for the bug in the PR, ie for something you will actually push!
Else skara and everyone else will get confused."

Thank you for your comment. I can close this PR and open a new one with a different number.

"Then there's the overall question as to how important, or appropriate is this API ?
Seems like its not really an A11Y API, its an arbitrary text-to-speech API.
I think of the A11Y APIs as being tied to the UI and making it accessible, not
providing some other way of communicating something which isn't even in the UI.
Put it this way, if I am NOT using an AT to provide speech, how are you communicating
whatever it is to the user - something changes in the UI, right ?
So then the A11Y API will be able to report that already so why do you need this ?
Show me some equivalent cases and uses in platform ATs APIs"

This functionality is part of the a11y API, for announcing we use the capabilities of the screen reader.
For example, NSAccessibility API capabilities for Mac: https://developer.apple.com/documentation/appkit/nsaccessibilitynotificationname?language=objc
Or NVDA controller client for Windows: https://github.com/nvaccess/nvda/tree/master/extras/controllerClient
Using the usual TTS annunciation tools is not a good idea, as the default TTS settings on the user's system can differ from those in the screen reader down to the voice, which is very confusing, and there may also be a difference in speed, volume and pace of speech.
Also, the built-in system TTS can sound simultaneously with the screen reader, it will sound like two voices speaking at the same time. Therefore, it will not be possible to implement interruption of the current output.

**Pull request closed without merging**